web/page0/src/app/runtime.service.ts

```diff
@@ -1,17 +1,225 @@
+  }
+
+  private getStoredRuntimes(): Runtime[] {
+    const runtimes = localStorage.getItem('runtimes');
+    // TODO: handle type error
+    if (runtimes) {
+      return JSON.parse(runtimes) as Runtime[];
+    }
+
+    return [];
+  }
+
+  private getInitRuntimeSettings() {
+    const storedRuntimes = this.getStoredRuntimes();
+    if (storedRuntimes.length === 0) {
+      return defaultRuntimeSettings;
+    }
+
+    return storedRuntimes.map((runtime) => ({
+      alias: runtime.alias,
+      url: runtime.url,
+    }));
+  }
+
+  private initRuntimeSettings = this.getInitRuntimeSettings();
+
+  private runtimes$: Observable<Runtime[]> = this.runtimeAction.pipe(
+    startWith({ type: 'init' } as RuntimeInitializeAction),
+    tap((action) => console.log(action)),
+    mergeScan((acc, action) => {
+      if (action.type === 'init') {
+        this.status$.next(RuntimesStatus.initializing);
+
+        return forkJoin(
+          this.initRuntimeSettings.map(({ alias, url }) =>
+            this.verifyRuntime(url, alias).pipe(
+              catchError((error) => {
+                console.error(error);
+                return of({
+                  url,
+                  alias,
+                  status: 'error',
+                } as Runtime);
+              })
+            )
+          )
+        ).pipe(
+          tap((runtimes) => {
+            this.status$.next(RuntimesStatus.done);
+          })
+        );
+      }
+
+      if (action.type === 'register') {
+        return of([...acc, action.runtime]);
+      }
+
+      if (action.type === 'unregister') {
+        return of(acc.filter((item) => item.alias !== action.alias));
+      }
+
+      if (action.type === 'refresh') {
+        this.status$.next(RuntimesStatus.refreshing);
+        return forkJoin(
+          acc.map((runtime) =>
+            this.verifyRuntime(runtime.url, runtime.alias).pipe(
+              catchError((error) =>
+                of({
+                  url: runtime.url,
+                  alias: runtime.alias,
+                  status: 'error',
+                } as Runtime)
+              )
+            )
+          )
+        ).pipe(
+          tap(() => {
+            this.status$.next(RuntimesStatus.done);
+          })
+        );
+      }
+      return of(acc);
+    }, [] as Runtime[]),
+    tap((runtimes) => console.log('runtimes', runtimes)),
+    tap((runtimes) =>
+      localStorage.setItem('runtimes', JSON.stringify(runtimes))
+    ),
+    shareReplay(1)
+  );
+
+  private status$ = new BehaviorSubject<RuntimesStatus>(
+    RuntimesStatus.initializing
+  );
+
+  getStatus() {
+    return this.status$;
+  }
 
   getRuntimes() {
-    return this.runtimes;
+    return this.runtimes$;
+  }
+
+  registerRuntime(alias: string, url: string) {
+    this.status$.next(RuntimesStatus.registering);
+
+    return this.checkDuplicatedAlias(alias).pipe(
+      mergeMap(() => this.verifyRuntime(url, alias)),
+      map((runtime) => this.register(runtime)),
```

Comment:
i think it'd better be `tap` instead `map` just in case that subscriber wants `runtime` data.

Reply:
Done